File: src/com/keunwon/object/chap02/Money.java
package com.keunwon.object.chap02;

import java.math.BigDecimal;

public class Money {
    public static final Money ZERO = Money.wons(0);

    private final BigDecimal amount;

    public static Money wons(long amount) {
        return new Money(BigDecimal.valueOf(amount));
    }

    public static Money wons(double amount) {
        return new Money(BigDecimal.valueOf(amount));
    }

    Money(BigDecimal amount) {
        this.amount = amount;
    }

    public Money plus(Money money) {
        return new Money(this.amount.add(money.amount));
    }

    public Money minus(Money money) {
        return new Money(this.amount.subtract(money.amount));
    }

    public Money times(double amount) {
        return new Money(this.amount.multiply(BigDecimal.valueOf(amount)));
    }

    public boolean isLessThan(Money other) {
        return amount.compareTo(other.amount) < 0;
    }

    public boolean isGreaterThanOrEqul(Money other) {
        return amount.compareTo(other.amount) >= 0;
    }

}

File: spring/Spring-JWT/core/src/main/kotlin/com/github/keunwon/core/generic/Money.kt
package com.github.keunwon.core.generic

import java.math.BigDecimal

class Money private constructor(val amount: BigDecimal) : Comparable<Money> {
    operator fun plus(amount: Money): Money = Money(this.amount.plus(amount.amount))

    operator fun minus(amount: Money): Money = Money(this.amount.minus(amount.amount))

    operator fun times(percent: Double): Money = Money(this.amount.times(BigDecimal.valueOf(percent)))

    operator fun div(divisor: Double): Money = Money(this.amount.div(BigDecimal.valueOf(divisor)))


    override fun compareTo(other: Money): Int {
        return this.amount.compareTo(other.amount)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Money

        if (amount != other.amount) return false

        return true
    }

    override fun hashCode(): Int {
        return amount.hashCode()
    }

    override fun toString(): String = "$amount 원"

    companion object {
        val ZERO = wons(0L)

        fun wons(amount: Long) = Money(BigDecimal.valueOf(amount))
        fun wons(amount: Double) = Money(BigDecimal.valueOf(amount))
        fun wons(amount: BigDecimal) = Money(amount)
    }
}
